iroh/src/magicsock/transports.rs

```diff
@@ -0,0 +1,434 @@
+        let relays = watcher::Join::new(self.relay.iter().map(|t| t.local_addr_watch()));
+
+        (ips, relays)
+            .map(|(ips, relays)| {
+                ips.into_iter()
+                    .map(Addr::from)
+                    .chain(
+                        relays
+                            .into_iter()
+                            .flatten()
+                            .map(|(relay_url, node_id)| Addr::RelayUrl(relay_url, node_id)),
+                    )
+                    .collect()
+            })
+            .expect("disconnected")
+    }
+
+    #[cfg(wasm_browser)]
+    pub(crate) fn local_addrs_watch(&self) -> impl Watcher<Value = Vec<Addr>> + Send + Sync {
+        let relays = self.relay.iter().map(|t| {
+            t.local_addr_watch()
+                .map(move |t| t.map(|(url, id)| Addr::RelayUrl(url, id)))
+                .expect("disconnected")
+        });
+        watcher::JoinOpt::new(relays)
+    }
+
+    /// Returns the bound addresses for IP based transports
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn ip_bind_addrs(&self) -> Vec<SocketAddr> {
+        self.ip.iter().map(|t| t.bind_addr()).collect()
+    }
+
+    /// Returns the bound addresses for IP based transports
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn ip_local_addrs(&self) -> Vec<SocketAddr> {
+        self.ip.iter().map(|t| t.local_addr()).collect()
+    }
+
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn max_transmit_segments(&self) -> usize {
+        let res = self.ip.iter().map(|t| t.max_transmit_segments()).min();
+        res.unwrap_or(1)
+    }
+
+    #[cfg(wasm_browser)]
+    pub(crate) fn max_transmit_segments(&self) -> usize {
+        1
+    }
+
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn max_receive_segments(&self) -> usize {
+        // `max_receive_segments` controls the size of the `RecvMeta` buffer
+        // that quinn creates. Having buffers slightly bigger than necessary
+        // isn't terrible, and makes sure a single socket can read the maximum
+        // amount with a single poll. We considered adding these numbers instead,
+        // but we never get data from both sockets at the same time in `poll_recv`
+        // and it's impossible and unnecessary to be refactored that way.
+
+        let res = self.ip.iter().map(|t| t.max_receive_segments()).max();
+        res.unwrap_or(1)
+    }
+
+    #[cfg(wasm_browser)]
+    pub(crate) fn max_receive_segments(&self) -> usize {
+        1
+    }
+
+    #[cfg(not(wasm_browser))]
+    pub(crate) fn may_fragment(&self) -> bool {
+        self.ip.iter().any(|t| t.may_fragment())
+    }
+
+    #[cfg(wasm_browser)]
+    pub(crate) fn may_fragment(&self) -> bool {
+        false
+    }
+
+    /// Check if a transport is writable, aka sendable on, for the given `addr`.
+    pub(crate) fn poll_writable(&self, cx: &mut Context, addr: &Addr) -> Poll<io::Result<()>> {
+        // TODO: what about multiple matches?
```

Comment:
Leftover TODO?
Yeah not sure. Perhaps worth experimenting with more than one relay transport or more than 2 IP transports, and seeing what happens here.
But tbh, the whole polling/writability logic is deeply flawed currently, so :man_shrugging: 
I really hope it'll get better when we eat more of the quinn code/support multiple sockets in quinn.

Reply:
Given the current setup it can't happen, as each transport supports exactly one `Addr` type, so I was planning on leaving this in for now